### PR TITLE
Allow setting Tx isolation on Tx start

### PR DESCRIPTION
**Pull request by dpetrick**
Allows setting an `IsolationLevel` when starting a new transaction. The underlying databases validate what levels of the ANSI + Snapshot levels are appropriate. Additionally, some databases require `SET` of the isolation level before a `BEGIN` of a transaction. 

Functional testing is thin, because it's hard to validate the isolation levels without either:
- Setting up dedicated scenarios for all databases to test.
- Checking that the connection executed the `SET` statement correctly.
- Somehow checking that the isolation level of the connection changed. Difficulty greatly depends on the database provider here. It's unclear if this is feasible for all providers.

### CODE REVIEW

File: src/tests/query.rs [hunk 1]
```diff
@@ -84,6 +84,60 @@ async fn transactions(api: &mut dyn TestApi) -> crate::Result<()> {
     Ok(())
 }
 
+#[test_each_connector(tags("mssql", "postgresql", "mysql"))]
+async fn transactions_with_isolation_works(api: &mut dyn TestApi) -> crate::Result<()> {
```
Comment (dpetrick) [inline src/tests/query.rs:88]: I don't have a great idea how to test that the TX isolation levels are set correctly, this is the best I came up with.

File: src/connector/mssql.rs [hunk 2]
```diff
@@ -412,6 +376,11 @@ impl Queryable for Mssql {
         self.query_raw(sql, params).await
     }
 
+    async fn execute(&self, q: Query<'_>) -> crate::Result<u64> {
+        let (sql, params) = visitor::Mssql::build(q)?;
```
Comment (garrensmith) [inline src/connector/mssql.rs:380]: Why was this moved?

File: src/connector/mssql.rs [hunk 3]
```diff
@@ -504,7 +484,12 @@ impl MssqlUrl {
         let transaction_isolation_level = props
             .remove("isolationlevel")
             .or_else(|| props.remove("isolation_level"))
-            .map(|level| IsolationLevel::from_str(&level))
+            .map(|level| {
```
Comment (garrensmith) [inline src/connector/mssql.rs:487]: This should be in its own smaller function.

**Review by pimeys — APPROVED** [review]

File: src/connector/mssql.rs [hunk 4]
```diff
@@ -87,56 +88,24 @@ pub(crate) struct MssqlQueryParams {
...
 #[async_trait]
 impl TransactionCapable for Mssql {
-    async fn start_transaction(&self) -> crate::Result<Transaction<'_>> {
-        Transaction::new(self, "BEGIN TRAN").await
+    async fn start_transaction(&self, isolation: Option<IsolationLevel>) -> crate::Result<Transaction<'_>> {
+        // Isolation levels in SQL Server are set on the connection and live until they're changed.
+        // Always explicitly setting the isolation level each time a tx is started (either to the given value
+        // or by using the default/connection string value) prevents transactions started on connections from
+        // the pool to have unexpected isolation levels set.
+        Transaction::new(
```
Comment (garrensmith) [inline src/connector/mssql.rs:100]: I would change this to something like:
```
Transaction::new(
  self,
  "BEGIN TRAN",
  self.get_isolation_level(isolation)
);

```
That way you can put all the isolation comments in that function.

File: src/connector/transaction.rs [hunk 5]
```diff
@@ -15,13 +19,31 @@ pub struct Transaction<'a> {
 }
 
 impl<'a> Transaction<'a> {
-    pub(crate) async fn new(inner: &'a dyn Queryable, begin_stmt: &str) -> crate::Result<Transaction<'a>> {
+    pub(crate) async fn new(
+        inner: &'a dyn Queryable,
+        begin_stmt: &str,
+        isolation_level: Option<IsolationLevel>,
+        isolation_first: bool,
```
[the remark this answers is not in the document]
Comment (Weakky) [reply src/connector/transaction.rs:26]: How about a `TransactionOptions` struct instead ?

**Review by Weakky — APPROVED** [review]
